### DISCUSSION TITLE
The root_path parameter of the FastAPI object doesn't work in "Behind a Proxy"

### DISCUSSION THREAD

**Discussion by insistence** [fastapi/fastapi#11033] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import uvicorn


app = FastAPI(
    root_path="/proxy"
)


@app.get("/app")
def read_main():
    return {"message": "Hello World from main app"}


subapi = FastAPI()


@subapi.get("/sub")
def read_sub():
    return {"message": "Hello World from sub API"}


app.mount("/subapi", subapi)


if __name__ == '__main__':
    uvicorn.run(app='app:app')
```

### Description

When I run the code, the server is expecting that I include the defined root path e.g:
http://localhost:8000/subapi/sub is not working (404)
http://localhost:8000/proxy/subapi/sub is working
IN version 0.108.0 It works as described here https://fastapi.tiangolo.com/advanced/behind-a-proxy/
My uvicorn version is 0.27.0
Thanks

### Operating System

Windows

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.9.18

**Comment by jgould22** [comment]
It is unclear to me to what the issue is specifically?

The behaviour you describe is expected behaviour.

When you set `root_path="/proxy"`  the entire application is mounted on that path.

```bash
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:56310 - "GET /proxy/app HTTP/1.1" 200 OK
INFO:     127.0.0.1:51872 - "GET /proxy/subapi/sub HTTP/1.1" 200 OK
```

**Comment by insistence** [comment]
> It is unclear to me to what the issue is specifically?
> 
> The behaviour you describe is expected behaviour.
> 
> When you set `root_path="/proxy"` the entire application is mounted on that path.
> 
> ```shell
> INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
> INFO:     127.0.0.1:56310 - "GET /proxy/app HTTP/1.1" 200 OK
> INFO:     127.0.0.1:51872 - "GET /proxy/subapi/sub HTTP/1.1" 200 OK
> ```

Perhaps there is an issue with my description. I am currently using a proxy like /dev-api in the Vue project to request static files under the fastapis service. I can retrieve files normally by specifying root-path as /dev-api using uvicorn, but only specifying the root_path parameter in the fastapis object cannot retrieve static files. I think this is a similar issue to #10948 and #10978. If my idea is incorrect, please correct it.

**Reply by insistence** [reply]
For the code I provided above, when I directly use uvicorn to specify root-path, I can access http://127.0.0.1:8000/subapi/sub in a browser. But when I only specify the root_path parameter in the FastAPI object, I cannot access http://127.0.0.1:8000/subapi/sub.

**Comment by fmelihh** [comment]
Hello @insistence , I saw that the problem was solved when you gave a root_path to the sub application you created. I hope the answer you want is this way. 
You can see the sample code below; 

```
import uvicorn
from fastapi import FastAPI


app = FastAPI(
    root_path="/proxy"
)


@app.get("/app")
def read_main():
    return {"message": "Hello World from main app"}


subapi = FastAPI(
    root_path="/subapi"
)


@subapi.get("/sub")
def read_sub():
    return {"message": "Hello World from sub API"}


app.mount("", subapi)


if __name__ == '__main__':
    uvicorn.run(app='app:app')
```

The operating system I use ubuntu. All other dependencies are the same as the dependencies you mentioned. If the solution you want is not this, I can try to help you if you can eleborate a bit more.

Happy coding :smiley:

**Reply by insistence** [reply]
Thanks for your reply. Your code can indeed solve the problem of sub application mounting, but if I mount staticfiles like the one below
```
import uvicorn
from fastapi import FastAPI
from fastapi.staticfiles import StaticFiles


app = FastAPI(
    root_path="/proxy"
)


@app.get("/app")
def read_main():
    return {"message": "Hello World from main app"}


app.mount("/profile", StaticFiles(directory="static"))


if __name__ == '__main__':
    uvicorn.run(app='app:app')
```
I already have a directory called static, which contains a file called test.txt. Currently, I can only access it by http://127.0.0.1:8000/proxy/profile/test.txt in a browser. I can't access it by http://127.0.0.1:8000/profile/test.txt in a browser. May I ask if there is a solution? Thank you very much.

**Reply by insistence** [reply]
In summary, I believe this is inconsistent with the description in the official documents of FastAPI（https://fastapi.tiangolo.com/advanced/behind-a-proxy/ ).  I'm not sure if I misunderstood the description of the document, or if it's a bug in FastAPI.

**Reply by fmelihh** [reply]
Hello @insistence yes I also think the document is not very clear. I found a solution for your problem on stackoverflow in this way. You can reach the relevant post from [here](https://stackoverflow.com/questions/62455652/how-to-serve-static-files-in-fastapi). 
I recommend you to download jinja2 package as dependency. 

You can find the updated code below; 
```
import uvicorn
from fastapi import FastAPI
from fastapi.requests import Request
from fastapi.templating import Jinja2Templates


app = FastAPI(root_path="/proxy")


@app.get("/app")
def read_main():
    return {"message": "Hello World from main app"}


subapi = FastAPI(root_path="/subapi")
templates = Jinja2Templates(directory="static")


@subapi.get("/sub")
def read_sub():
    return {"message": "Hello World from sub API"}


@subapi.get("/items/test_txt")
def example(request: Request):
    return templates.TemplateResponse("test.txt", {"request": request})


app.mount("", subapi)

if __name__ == "__main__":
    uvicorn.run(app="main:app")

```

**Reply by insistence** [reply]
Thank you very much for your answer. This is possible in certain scenarios, but it seems to not meet my needs as the file path and file name of my current project are dynamically generated. The StaticFiles of FastAPI meet my needs, except for setting the root_path parameter in FastAPI which cannot take effect, everything else is fine. I think we should ask @tiangolo to confirm if this is a bug. Thank you again.

**Reply by insistence** [reply]
@tiangolo Can you spare some time to help take a look at this question? Thank you very much!

**Comment by R3myG** [comment]
Just adding an additional datapoint here.
Upgraded from 0.104.1 to 0.109.0 and suddenly the the following code:

```
app = FastAPI(root_path="/api")

app.mount("/reference", staticfiles.StaticFiles(directory="/api/ref/site/", html=True), name="reference")
```
can only be accessed via the `/api/api/reference` url while in 0.104.1 it's accessible as expected at `/api/reference`.

Downgrading back to 0.104.1 resolves the issue. 

Note that I've tried to upgrade to 0.109.2 that was recently released and the same issue is occurring as with 0.109.0

**Reply by bertomaniac** [reply]
@R3myG 
I can confirm the same problem.

```python
fast_app = FastAPI(
        title="My Service",
        version=__version__,
        root_path="/some/root/path",
        docs_url=None,
        redoc_url=None,
)

fast_app.mount("/static", StaticFiles(directory=settings.static_dir), name="static")
```

Self-hosted docs
```python
    @router.get("/", include_in_schema=False)
    async def custom_swagger_ui_html() -> HTMLResponse:
        """Set up a localized version of the Swagger (OpenAPI) docs that can be run without assets from the Internet."""

        return get_swagger_ui_html(
            openapi_url="/openapi.json",
            title="My Service",
            swagger_js_url="/static/swagger-ui-bundle.js",
            swagger_css_url="/static/swagger-ui.css",
            swagger_favicon_url="/static/favicon.ico",
            swagger_ui_parameters={
                "queryConfigEnabled": False,  # prevents overriding settings in url string
                # "supportedSubmitMethods": [],  # disables interactivity
                "defaultModelRendering": "model",
                # "docExpansion": "none",
                "syntaxHighlight": True,
            },
        )
```

Test for the docs:
```python
def test_get_docs(client: TestClient):
    response = client.get("/")
    assert response.status_code == 200
    assert response.headers["Content-Type"] == "text/html; charset=utf-8"

    response = client.get("/openapi.json")
    assert response.status_code == 200

    response = client.get("/static/swagger-ui-bundle.js")
    assert response.status_code == 200

    response = client.get("/static/swagger-ui.css")
    assert response.status_code == 200

    response = client.get("/static/favicon.ico")
    assert response.status_code == 200
```

The above test works in `0.104.1` and does not work when I upgrade to `0.110.0`

**Reply by yinziyan1206** [reply]
Are you using gunicorn

**Reply by bertomaniac** [reply]
No, uvicorn

**Reply by hanak** [reply]
Thank you very much for providing the version number. It saved the rest of the day for me.

**Comment by richin13** [comment]
I think I might've fixed this in https://github.com/tiangolo/fastapi/pull/11160

I was having an issue with the 307 redirect that happens when you don't include the trailing slash, when running behind a proxy the `Location` header of the redirect did not include the `root_path` due to an inconsistency of the scope data going around internally

**Reply by ed9w2in6** [reply]
I have NOT read your fix.

However, I have been fiddling with `root_path` for my `pytest` test cases and noticed that setting `root_path` to `"/"` will now fail every requests in which setting to `""` fixes it.
